homeassistant/components/hunterdouglas_powerview/cover.py

```diff
@@ -341,7 +411,189 @@ class PowerViewShade(PowerViewShadeBase):
+        """Return if the cover is closed."""
+        # treat anything below 75% of 1% of total position as closed due to conversion of powerview to hass
+        return self._current_hd_cover_secondary <= CLOSED_POSITION
+
+    async def async_open_cover(self, **kwargs):
+        """Open the cover."""
+        current_hass_pos1 = self.current_cover_position
+        current_hass_pos2 = self.current_cover_position_secondary
+        steps_to_move = 100 - (current_hass_pos1 + current_hass_pos2)
+        self._async_schedule_update_for_transition(steps_to_move)
+        self._async_update_from_command(
+            await self._shade.move(
+                {
+                    ATTR_POSITION1: MIN_POSITION,
+                    ATTR_POSITION2: MAX_POSITION,
+                    ATTR_POSKIND1: POSKIND_PRIMARY,
+                    ATTR_POSKIND2: POSKIND_SECONDARY,
+                }
+            )
+        )
+
+    @property
+    def current_cover_position(self):
+        """Return the current position of cover."""
+        # these need to be inverted to report state properly in HA
+        return hd_position_to_hass(self._current_hd_cover_secondary)
+
+    @property
+    def current_cover_position_bottom(self):
+        """Return the current position of cover."""
+        # these need to be inverted to report state properly in HA
+        return hd_position_to_hass(self._current_hd_cover_primary)
+
+    async def _async_move(self, target_hass_position):
+        """Move the shade to a position."""
+        # dont allow a cover to go past the position of the opposite motor
+        cover_bottom = self.current_cover_position_bottom
+        if (100 - target_hass_position) < cover_bottom:
+            target_hass_position = (100 - cover_bottom) - 1
+
+        current_hass_position = self.current_cover_position
+        steps_to_move = abs(current_hass_position - target_hass_position)
+        self._async_schedule_update_for_transition(steps_to_move)
+
+        motor_position_bottom = self.get_position_primary
+        motor_position_top = hass_position_to_hd(target_hass_position)
+
+        if self.coordinator.data is not None:
+            self.set_position_secondary(hass_position_to_hd(target_hass_position))
+            _LOGGER.debug(self.coordinator.data[self._shade.id][ATTR_POSITION_DATA])
+
+        self._async_update_from_command(
+            await self._shade.move(
+                {
+                    ATTR_POSITION1: motor_position_bottom,
+                    ATTR_POSITION2: motor_position_top,
+                    ATTR_POSKIND1: POSKIND_PRIMARY,
+                    ATTR_POSKIND2: POSKIND_SECONDARY,
+                }
+            )
+        )
+        self._is_opening = False
+        self._is_closing = False
+        if target_hass_position > current_hass_position:
+            self._is_opening = True
+        elif target_hass_position < current_hass_position:
+            self._is_closing = True
+        self.async_write_ha_state()
```

Comment:
Can you make this more DRY?  Its very similar to the code in `PowerViewShadeTDBUBottom`

Please see comment in `PowerViewShadeTDBUBottom` as well

Reply:
Cleaned this up a lot - including the base class one

Not sure how to get around line 573/574 above as each call (PowerViewShadeTDBUBottom vs PowerViewShadeTDBUTop) require different values

So i could pull them into their own functions and move to 1 async_move under PowerViewShadeTDBU but this could be more confusing than the current setup ?

Reply:
_async_move is now only defined once. Please take another look

The new DAtaCoordinator seems to be working 'OK' it is not instantaneous and i am trying to work out if that is becase of the _async_update_from_command command or what

there appear to be a couple of times data is written to the coordinator after a move command (not even including the resync that is done)

not sure if you have any suggestions to try and get the two shades more in sync - It works fine as long as you arent spamming commands - somewhere in there the details in position1/position2 seem to get overwritten